src/stash/src/transaction.rs

```diff
@@ -558,41 +559,74 @@ impl<'a> Transaction<'a> {
-                    )
-                    .map_err(|err| err.into())
-                    .await
-            });
-        }
-
         // Check the upper in a separate future so we can issue the updates without
         // waiting for it first.
-        try_join(
-            async {
-                let upper = match upper {
-                    Some(upper) => upper,
-                    None => self.upper(collection_id).await?,
-                };
-                for ((_key, _value), time, _diff) in entries {
-                    if !upper.less_equal(time) {
-                        return Err(StashError::from(format!(
-                            "entry time {} is less than the current upper frontier {}",
-                            time,
-                            AntichainFormatter(&upper)
-                        )));
-                    }
+        let upper_fut = async {
+            let upper = match upper {
+                Some(upper) => upper,
+                None => self.upper(collection_id).await?,
+            };
+            for ((_key, _value), time, _diff) in entries {
+                if !upper.less_equal(time) {
+                    return Err(StashError::from(format!(
+                        "entry time {} is less than the current upper frontier {}",
+                        time,
+                        AntichainFormatter(&upper)
+                    )));
                 }
-                Ok(upper)
-            },
-            try_join_all(futures),
-        )
-        .await?;
+            }
+            Ok(upper)
+        };
+
+        // For larger update sets, generate a dynamic statement with all the
+        // arguments. We don't always do this because using the existing
+        // prepared statement avoids a few roundtrips to the server preparing
+        // the new one.
+        //
+        // TODO: 2 is made up. We should benchmark the tradeoff point sometime.
+        if entries.len() > 2 {
+            let mut stmt =
+                String::from("INSERT INTO data (collection_id, key, value, time, diff) VALUES");
+            let mut sep = ' ';
+            let mut args: Vec<&'_ (dyn ToSql + Sync)> = Vec::with_capacity(1 + entries.len() * 4);
+            // All rows use the collection id, so hard code it as the first.
+            args.push(&collection_id);
+            for (i, ((key, value), time, diff)) in entries.into_iter().enumerate() {
+                let idx = 1 + i * 4;
+                args.push(key);
+                args.push(value);
+                args.push(time);
+                args.push(diff);
+                write!(
+                    &mut stmt,
+                    "{}($1, ${}, ${}, ${}, ${})",
+                    sep,
+                    idx + 1,
+                    idx + 2,
+                    idx + 3,
+                    idx + 4
+                )
+                .unwrap();
+                sep = ',';
```

Comment:
This is 100% subjective, but it might be slightly easier to read if this statement building logic was in it's own function.

Reply:
Agreed! I did a refactor and moved this into the PreparedStatements struct so it's properly accounted too.